pkg/handler/iam_handler.go

```diff
@@ -0,0 +1,209 @@
+
+// IAMHandler updates IAM policies of GCP organizations, folders, and projects
+// based on the IAM request received.
+type IAMHandler struct {
+	organizationsClient *resourcemanager.OrganizationsClient
+	foldersClient       *resourcemanager.FoldersClient
+	projectsClient      *resourcemanager.ProjectsClient
+	// Optional retry backoff strategy, default is 5 attempts with fibonacci
+	// backoff that starts at 500ms.
+	retry retry.Backoff
+}
+
+// Internal iamClient interface that get and set IAM policies for GCP
+// organizations, folders, and projects.
+type iamClient interface {
+	GetIamPolicy(context.Context, *iampb.GetIamPolicyRequest, ...gax.CallOption) (*iampb.Policy, error)
+	SetIamPolicy(context.Context, *iampb.SetIamPolicyRequest, ...gax.CallOption) (*iampb.Policy, error)
+}
+
+// Option is the option to set up an IAMHandler.
+type Option func(h *IAMHandler) (*IAMHandler, error)
+
+// WithRetry provides retry strategy to the handler.
+func WithRetry(b retry.Backoff) Option {
+	return func(p *IAMHandler) (*IAMHandler, error) {
+		p.retry = b
+		return p, nil
+	}
+}
+
+// NewIAMHandler creates a new IAMHandler with provided clients and options.
+func NewIAMHandler(ctx context.Context, o *resourcemanager.OrganizationsClient, f *resourcemanager.FoldersClient, p *resourcemanager.ProjectsClient, opts ...Option) (*IAMHandler, error) {
+	h := &IAMHandler{}
+	for _, opt := range opts {
+		var err error
+		h, err = opt(h)
+		if err != nil {
+			return nil, fmt.Errorf("failed to apply client options: %w", err)
+		}
+	}
+	h.organizationsClient = o
+	h.foldersClient = f
+	h.projectsClient = p
+
+	if h.retry == nil {
+		h.retry = retry.WithMaxRetries(5, retry.NewFibonacci(500*time.Millisecond))
+	}
+	return h, nil
+}
+
+// Do removes expired or duplicative IAM bindings added by AOD and adds requested IAM bindings to current IAM policy.
+func (h *IAMHandler) Do(ctx context.Context, r *v1alpha1.IAMRequestWrapper) (nps []*v1alpha1.IAMResponse, retErr error) {
+	for _, p := range r.ResourcePolicies {
+		np, err := h.handlePolicy(ctx, p, r.Duration)
+		if err != nil {
+			retErr = errors.Join(
+				retErr,
+				fmt.Errorf("failed to handle policy update for resource %s: %w", p.Resource, err),
+			)
+		}
+		if np != nil {
+			nps = append(nps, np)
+		}
+	}
+	return
+}
+
+func (h *IAMHandler) handlePolicy(ctx context.Context, p *v1alpha1.ResourcePolicy, ttl time.Duration) (*v1alpha1.IAMResponse, error) {
+	var iamC iamClient
+	switch strings.Split(p.Resource, "/")[0] {
+	case "organizations":
+		iamC = h.organizationsClient
+	case "folders":
+		iamC = h.foldersClient
+	case "projects":
+		iamC = h.projectsClient
+	default:
+		return nil, fmt.Errorf("unable to match none of the organizations, folders, projects resource")
+	}
+
+	var np *iampb.Policy
+	if err := retry.Do(ctx, h.retry, func(ctx context.Context) error {
+		// Get current IAM policy.
+		cp, err := iamC.GetIamPolicy(
+			ctx,
+			&iampb.GetIamPolicyRequest{Resource: p.Resource})
```

Comment:
nit: seems unnecessary line breaks?
```suggestion
		cp, err := iamC.GetIamPolicy(ctx, &iampb.GetIamPolicyRequest{Resource: p.Resource})
```